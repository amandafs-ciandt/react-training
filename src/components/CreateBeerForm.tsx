import React, { useState, useEffect } from 'react';
import { Input, Checkbox, Button, Select, Form } from 'antd';
import { submitForm } from '../effector/event';
import { Store } from 'antd/lib/form/interface';

const { Option } = Select;

const layout = {
  labelCol: {
    sm: { span: 5 },
  },
  wrapperCol: {
    sm: { span: 16 },
  },
};

const tailLayout = {
  wrapperCol: {
    sm: {
      span: 16,
      offset: 5,
    },
  },
};

const initialValues = {
  beerName: '',
  beerType: '',
  hasCorn: false,
  ingredients: '',
};

const CreateBeerForm = ({ notify }) => {
  const [form] = Form.useForm();

  const [validForm, setValidForm] = useState(false);

  useEffect(() => {
    setValidForm(
      form.getFieldValue('beerName').length > 0 &&
        form.getFieldValue('beerType').length > 0 &&
        form.getFieldValue('ingredients').length > 0
    );
  }, [form]);

  const onSubmit = (values: any): void => {
    submitForm(values);

    notify('Form successfully submitted!');
    form.resetFields();
  };

  return (
    <Form form={form} initialValues={initialValues} onFinish={onSubmit}>
      <Form.Item
        {...layout}
        label='Beer name'
        name='beerName'
        rules={[{ required: true, message: 'Please input the beer name!' }]}>
        <Input />
      </Form.Item>

      <Form.Item
        {...layout}
        name='beerType'
        label='Beer type'
        rules={[{ required: true, message: 'Please input the beer type!' }]}>
        <Select placeholder='Select a beer type'>
          <Option value='ale'>Ale</Option>
          <Option value='lager'>Lager</Option>
          <Option value='malt'>Malt</Option>
          <Option value='stout'>Stout</Option>
        </Select>
      </Form.Item>

      <Form.Item {...tailLayout} name='hasCorn' valuePropName='checked'>
        <Checkbox>Has corn</Checkbox>
      </Form.Item>

      <Form.Item
        {...layout}
        name='ingredients'
        label='Ingredients'
        rules={[{ required: true, message: 'Please input the ingredients!' }]}>
        <Input.TextArea />
      </Form.Item>

      <Form.Item {...tailLayout}>
        <Button type='primary' htmlType='submit' disabled={!validForm}>
          Submit
        </Button>
      </Form.Item>
    </Form>
  );
};

export default CreateBeerForm;
