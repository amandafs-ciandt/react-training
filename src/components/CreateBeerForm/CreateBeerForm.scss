.create-beer-form {
  display: flex;
  flex-direction: column;
  padding: 2rem;
  padding-top: 0;

  .margin-bottom-one {
    margin-bottom: 1rem;
  }
}
