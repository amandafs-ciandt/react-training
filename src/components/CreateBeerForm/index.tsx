import React, { useState, useEffect, FormEvent, ChangeEvent } from 'react';
import {
  TextField,
  FormControl,
  InputLabel,
  Select,
  MenuItem,
  FormControlLabel,
  Checkbox,
  Button,
} from '@material-ui/core';

import './CreateBeerForm.scss';

interface IProps {
  notify: Function;
}

type FormData = {
  beerName: string;
  beerType: string;
  hasCorn: boolean;
  ingredients: string;
};

const defaultValues = {
  beerName: '',
  beerType: '',
  hasCorn: false,
  ingredients: '',
};

const CreateBeerForm = ({ notify }: IProps) => {
  const [form, setForm] = useState<FormData>(defaultValues);
  const [validForm, setValidForm] = useState(false);

  const onInputChange = (
    event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    const { name, value } = event.target;
    setForm({ ...form, [name]: value });
  };

  const onSelectChange = (event: ChangeEvent<any>) => {
    setForm({ ...form, [event.target.name]: event.target.value });
  };

  const onCheckboxChange = (event: ChangeEvent<any>) => {
    setForm({ ...form, [event.target.name]: event.target.checked });
  };

  const onSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    notify();
  };

  useEffect(() => {
    setValidForm(
      form.beerName.length > 0 &&
        form.beerType.length > 0 &&
        form.ingredients.length > 0
    );
  }, [form]);

  return (
    <div className='create-beer-form'>
      <h1 className='create-beer-form__title'>Beer Form</h1>
      <form
        className='create-beer-form__form'
        onSubmit={onSubmit}
        data-testid='form'>
        <div className='create-beer-form__input-container'>
          <TextField
            label='Beer name'
            id='beerName'
            name='beerName'
            variant='outlined'
            onChange={onInputChange}
            value={form.beerName}
          />
        </div>
        <div className='create-beer-form__input-container'>
          <FormControl variant='outlined'>
            <InputLabel id='beerTypeLabel'>Beer type</InputLabel>
            <Select
              labelId='beerTypeLabel'
              id='beerType'
              label='Beer type'
              onChange={onSelectChange}
              value={form.beerType}
              inputProps={{
                name: 'beerType',
                id: 'beerType',
              }}>
              <MenuItem data-testid='beer-type-option' value='ale'>
                Ale
              </MenuItem>
              <MenuItem data-testid='beer-type-option' value='lager'>
                Lager
              </MenuItem>
              <MenuItem data-testid='beer-type-option' value='malt'>
                Malt
              </MenuItem>
              <MenuItem data-testid='beer-type-option' value='stout'>
                Stout
              </MenuItem>
            </Select>
          </FormControl>
        </div>
        <div className='create-beer-form__input-container'>
          <FormControlLabel
            control={<Checkbox color='primary' />}
            label='Has corn'
            labelPlacement='end'
            name='hasCorn'
            id='hasCorn'
            onChange={onCheckboxChange}
            checked={form.hasCorn}
          />
        </div>
        <div className='create-beer-form__input-container'>
          <TextField
            id='ingredients'
            label='Ingredients'
            multiline
            rows={3}
            variant='outlined'
            name='ingredients'
            onChange={onInputChange}
            value={form.ingredients}
          />
        </div>
        <Button
          data-testid='form-submit-button'
          type='submit'
          variant='contained'
          color='secondary'
          disabled={!validForm}>
          Submit
        </Button>
      </form>
    </div>
  );
};

export default CreateBeerForm;
