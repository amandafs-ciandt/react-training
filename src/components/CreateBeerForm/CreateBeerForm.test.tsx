import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import CreateBeerForm from '.';

const mockedNotify = jest.fn();

const setUpShallowRendering = (): ShallowWrapper => {
  return shallow(<CreateBeerForm notify={mockedNotify} />);
};

describe('CreateBeerForm', () => {
  let createBeerFormComponent: ShallowWrapper;

  beforeEach(() => {
    createBeerFormComponent = setUpShallowRendering();
  });

  it('should render correctly', () => {
    expect(createBeerFormComponent).toMatchSnapshot();
  });

  it('should render title', () => {
    const expectedTitle = 'Beer Form';
    const title = createBeerFormComponent
      .find('.create-beer-form__title')
      .text();

    expect(title).toBe(expectedTitle);
  });

  it('should render the form', () => {
    const formExists = createBeerFormComponent.exists(
      '.create-beer-form__form'
    );

    expect(formExists).toBeTruthy();
  });

  it('should render the inputs', () => {
    const numberOfInputs = 4;
    const formInputs = createBeerFormComponent.find(
      '.create-beer-form__input-container'
    );

    expect(formInputs.length).toEqual(numberOfInputs);
  });

  it('should have submit button disabled after form is rendered', () => {
    const submitButton = createBeerFormComponent.find(
      '[data-testid="form-submit-button"]'
    );

    expect(submitButton.props().disabled).toBeTruthy();
  });

  it('should have submit button disabled when form is not valid', () => {
    createBeerFormComponent.find('#beerType').simulate('change', {
      persist: () => {},
      target: {
        name: 'beerName',
        value: 'Budweiser',
      },
    });

    const submitButton = createBeerFormComponent.find(
      '[data-testid="form-submit-button"]'
    );

    expect(submitButton.props().disabled).toBeTruthy();
  });

  it('should have submit button disabled when form is valid', () => {
    //TODO
  });

  it('should update beerName when it is changed', () => {
    createBeerFormComponent.find('#beerName').simulate('change', {
      persist: () => {},
      target: {
        name: 'beerName',
        value: 'Budweiser',
      },
    });

    const newValue = createBeerFormComponent.find('#beerName').props().value;

    expect(newValue).toEqual('Budweiser');
  });

  it('should update beerType when it is changed', () => {
    createBeerFormComponent.find('#beerType').simulate('change', {
      persist: () => {},
      target: {
        name: 'beerType',
        value: 'lager',
      },
    });

    const newValue = createBeerFormComponent.find('#beerType').props().value;

    expect(newValue).toEqual('lager');
  });

  it('should update hasCorn when it is changed', () => {
    createBeerFormComponent.find('#hasCorn').simulate('change', {
      persist: () => {},
      target: {
        name: 'hasCorn',
        checked: true,
      },
    });

    const newValue = createBeerFormComponent.find('#hasCorn').props().checked;

    expect(newValue).toEqual(true);
  });

  it('should update ingredients when it is changed', () => {
    const ingredientsValue = 'Water, barley malt, rice, yeast and hops';

    createBeerFormComponent.find('#ingredients').simulate('change', {
      persist: () => {},
      target: {
        name: 'ingredients',
        value: ingredientsValue,
      },
    });

    const newValue = createBeerFormComponent.find('#ingredients').props().value;

    expect(newValue).toEqual(ingredientsValue);
  });

  it('should submit form correctly', () => {
    createBeerFormComponent
      .find('form')
      .simulate('submit', { preventDefault() {} });

    expect(mockedNotify).toHaveBeenCalledTimes(1);
  });

  it('should show the correct beer types on select', () => {
    const beerTypes = ['ale', 'lager', 'malt', 'stout'];
    const optionElements = createBeerFormComponent.find(
      '[data-testid="beer-type-option"]'
    );

    expect(optionElements.length).toEqual(beerTypes.length);

    expect(optionElements.get(0).props.value).toEqual(beerTypes[0]);
    expect(optionElements.get(1).props.value).toEqual(beerTypes[1]);
    expect(optionElements.get(2).props.value).toEqual(beerTypes[2]);
    expect(optionElements.get(3).props.value).toEqual(beerTypes[3]);
  });
});
