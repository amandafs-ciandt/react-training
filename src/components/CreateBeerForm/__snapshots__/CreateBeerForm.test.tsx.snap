// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateBeerForm should render correctly 1`] = `ShallowWrapper {}`;
